File: LLD/Design-Patterns/Strategy/code/strategy.py
"""
Strategy Pattern — a ShoppingCart (Context) delegates payment to a swappable PaymentStrategy.
Run: python strategy.py
"""

from abc import ABC, abstractmethod


class PaymentStrategy(ABC):
    @abstractmethod
    def pay(self, amount):
        ...


class CreditCardPayment(PaymentStrategy):
    def __init__(self, card_number):
        self.card_number = card_number

    def pay(self, amount):
        last4 = self.card_number[-4:]
        print(f"Charged ${amount} to credit card ending in {last4}")


class PayPalPayment(PaymentStrategy):
    def __init__(self, email):
        self.email = email

    def pay(self, amount):
        print(f"Charged ${amount} via PayPal account {self.email}")


class StoreCreditPayment(PaymentStrategy):
    def __init__(self, balance):
        self.balance = balance

    def pay(self, amount):
        if amount > self.balance:
            print(f"Store credit insufficient: have ${self.balance}, need ${amount}")
            return
        self.balance -= amount
        print(f"Charged ${amount} to store credit, remaining balance ${self.balance}")


class ShoppingCart:
    def __init__(self):
        self.payment_strategy = None

    def set_payment_strategy(self, strategy: PaymentStrategy):
        self.payment_strategy = strategy

    def checkout(self, amount):
        self.payment_strategy.pay(amount)


if __name__ == "__main__":
    cart = ShoppingCart()

    cart.set_payment_strategy(CreditCardPayment("4111111111111234"))
    cart.checkout(49.99)

    cart.set_payment_strategy(PayPalPayment("shopper@example.com"))
    cart.checkout(19.50)

    cart.set_payment_strategy(StoreCreditPayment(30.0))
    cart.checkout(45.0)
    cart.checkout(20.0)

File: v1/LLD/Design-Patterns/Strategy/code/Strategy.java
// Strategy Pattern — a ShoppingCart (Context) delegates payment to a swappable PaymentStrategy.
// Compile: javac Strategy.java
// Run:     java Strategy

interface PaymentStrategy {
    void pay(double amount);
}

class CreditCardPayment implements PaymentStrategy {
    private String cardNumber;

    CreditCardPayment(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    public void pay(double amount) {
        String last4 = cardNumber.substring(cardNumber.length() - 4);
        System.out.println("Charged $" + amount + " to credit card ending in " + last4);
    }
}

class PayPalPayment implements PaymentStrategy {
    private String email;

    PayPalPayment(String email) {
        this.email = email;
    }

    public void pay(double amount) {
        System.out.println("Charged $" + amount + " via PayPal account " + email);
    }
}

class StoreCreditPayment implements PaymentStrategy {
    private double balance;

    StoreCreditPayment(double balance) {
        this.balance = balance;
    }

    public void pay(double amount) {
        if (amount > balance) {
            System.out.println("Store credit insufficient: have $" + balance + ", need $" + amount);
            return;
        }
        balance -= amount;
        System.out.println("Charged $" + amount + " to store credit, remaining balance $" + balance);
    }
}

class ShoppingCart {
    private PaymentStrategy paymentStrategy;

    void setPaymentStrategy(PaymentStrategy strategy) {
        this.paymentStrategy = strategy;
    }

    void checkout(double amount) {
        paymentStrategy.pay(amount);
    }
}

public class Strategy {
    public static void main(String[] args) {
        ShoppingCart cart = new ShoppingCart();

        cart.setPaymentStrategy(new CreditCardPayment("4111111111111234"));
        cart.checkout(49.99);

        cart.setPaymentStrategy(new PayPalPayment("shopper@example.com"));
        cart.checkout(19.50);

        cart.setPaymentStrategy(new StoreCreditPayment(30.0));
        cart.checkout(45.0);
        cart.checkout(20.0);
    }
}
